### ISSUE TITLE
Make CPU softmax distinguish invalid logits from fully masked rows and reject malformed shapes

### ISSUE THREAD

**Issue by forkwright** [forkwright/logismos#169]
## Finding

**Priority: P2 — numerical integrity of production attention and CPU references.** Both CPU softmax implementations misclassify an all-NaN row as a fully masked row and return a finite uniform distribution. The fp32 path additionally retains a release-only silent zero-fill path for malformed shapes.

Verified by source/control-flow analysis at **`bb1e7468c3397d2c74ec27f5798d69bb9df3bd81`**. The Rust witnesses below were not compiled in the audit environment; no GPU or model execution is claimed.

## Evidence

- [`crates/kernels/src/cpu_f32/mod.rs:614-667`](https://github.com/forkwright/logismos/blob/bb1e7468c3397d2c74ec27f5798d69bb9df3bd81/crates/kernels/src/cpu_f32/mod.rs#L614-L667) initializes the row maximum to negative infinity and updates it only when `v > m`. Every comparison with NaN is false. Consequently `[NaN, NaN]`, or `[NaN, -inf]`, leaves `m == -inf` and enters the uniform-output branch at lines 629-649. The branch's assertion in prose that every element is `-inf` is not established by that condition.
- [`crates/kernels/src/softmax/cpu.rs:14-74`](https://github.com/forkwright/logismos/blob/bb1e7468c3397d2c74ec27f5798d69bb9df3bd81/crates/kernels/src/softmax/cpu.rs#L14-L74) repeats the same classification in `softmax_fp16_ref`, returning **`Ok`** for the fabricated probabilities despite already having a fallible API.
- The fp32 path checks `x.len() == rows * n` only with `debug_assert_eq!`, then uses `get(...).else { continue }`. With debug assertions disabled, a short input leaves the missing output row zero-filled and returns success. Both shape products also use unchecked multiplication.
- This is reachable production code: `crates/transformers/src/attention.rs:278` and `crates/transformers/src/modernbert.rs:430` call `cpu_f32::softmax_last_dim` during attention.

## Minimal witnesses

Current source behavior, expressed as Rust regression witnesses:

```rust
// Both debug and release: invalid numerical input becomes plausible probabilities.
let p = kernels::cpu_f32::softmax_last_dim(&[f32::NAN, f32::NAN], 1, 2);
assert_eq!(p, vec![0.5, 0.5]);

// Release only: the missing second input row becomes a successful zero output row.
let p = kernels::cpu_f32::softmax_last_dim(&[0.0, 0.0], 2, 2);
assert_eq!(p, vec![0.5, 0.5, 0.0, 0.0]);
```

The first witness follows directly from the comparison and branch conditions; it does not depend on a release optimization or uncommon allocation behavior. Add the equivalent fp16 witness against the existing fallible reference API.

## Why this matters

A CPU reference must expose invalid state, not manufacture a numerically plausible answer that can become a GPU parity target. Returning a uniform distribution masks upstream numerical failure and directs attention to data despite there being no valid logits. Debug-only shape checks also give the same public operation different correctness contracts in debug and release.

This is not a request to silently change the intentionally chosen all-`-inf` policy. Legitimate fully masked input, invalid numerical input, and malformed storage are distinct states and must be distinguishable first.

## Desired correction

Establish one checked domain contract for the production CPU softmax boundaries: checked dimensions/products, exact slice lengths, an explicit empty-axis policy, and explicit NaN/+infinity handling. Preserve any permitted `-inf` masking semantics only after proving that the row belongs to that domain. Use typed errors and migrate production callers; do not replace silent success with an unconditional library panic or another sentinel distribution.

Keep independent numerical expectations independent. Share boundary validation where appropriate, not a production implementation with its purported oracle.

## Done when

- All-NaN, mixed NaN/`-inf`, mixed finite/NaN, and positive-infinity rows cannot return successful finite probabilities by accident.
- Exact fully masked rows exercise a separately documented and tested policy.
- Short/long input, multiplication overflow and zero-width cases have the same typed behavior in debug and release.
- The fp32 attention callers propagate failure through their existing error boundaries.
- Adversarial tests run in both debug and release within the GPU-denied lane; genuine finite-input numerical parity is retained.

## Related work

#30 and #59 repaired the previous all-`-inf` propagation defect; the present finding concerns the replacement branch's incomplete classification and the still-unchecked fp32 shape boundary. #142 owns ModernBERT's separate head-layout/mask duplication. This issue does not duplicate that work or claim native Qwen decoder execution uses this legacy softmax.
